architecture.tex

```diff
@@ -53,58 +55,190 @@
+instance \texttt{Symbol('x', positive=True)} will create a symbol named
+\texttt{x} that is assumed to be positive.
+
+\begin{verbatim}
+>>> x = Symbol('x', positive=True)
+>>> sqrt(x**2)
+x
+\end{verbatim}
+
+Some common assumptions that SymPy allows are \texttt{positive},
+\texttt{negative}, \texttt{real}, \texttt{nonpositive}, \texttt{nonnegative},
+\texttt{real}, \texttt{integer}, and \texttt{commutative}~\footnote{If $A$ and
+  $B$ are Symbols created with \texttt{commutative=False} then SymPy will keep
+  $AB$ and $BA$ distinct.}. Assumptions on any object can be checked with the
+\verb|is_|\texttt{\textit{assumption}} attributes, like \verb|x.is_positive|.
+
+Assumptions are needed only to restrict a domain so that certain
+simplifications can be performed. It is not required to make the domain match
+the input of a function. For instance, one can create the object
+$\sum_{n=0}^m f(n)$ as \texttt{Sum(f(n), (n, 0, m))} without setting
+\texttt{integer=True} when creating the Symbol object \texttt{n}.
+
+The assumptions system additionally has deductive capabilities. The
+assumptions use a three-valued logic using the Python builtin objects
+\texttt{True}, \texttt{False}, and \texttt{None}. \texttt{None} represents the
+``unknown'' case, meaning either the given assumption can be either true or
+false under the given information, for instance, \verb|Symbol('x', real=True).is_real|
+will give \texttt{None} because a real symbol might be
+positive or it might not, or that not enough is implemented to determine, for
+instance, \verb|(pi + E).is_irrational| gives \texttt{None}, because SymPy
+does not know how to determine if $\pi + e$ is rational or irrational.
+
+Basic implications between the facts are used to deduce assumptions. For
+instance, the assumptions system knows that being an integer implies being
+rational, so \verb|Symbol('x', integer=True).is_rational| returns
+\texttt{True}. Furthermore, expressions compute the assumptions on themselves
+based on the assumptions of their arguments. For instance, if \texttt{x} and
+\texttt{y} are both created with \texttt{positive=True}, then \verb|(x + y).is_positive|
+will be \texttt{True}.
+
+SymPy also has an experimental assumptions system where facts are stored
+separate from objects, and deductions are made with a SAT solver. We will not
+discuss this system here.
+
 %%
 %% Extensibility
 \subsection{Extensibility}
 
-%% Functions
-\subsection{Functions}
+Extensibility is an important feature for SymPy. Because the same language,
+Python, is used both for the internal implementation and the external usage by
+users, all the extensibility capabilities available to users are also used by
+functions that are part of SymPy.
+
+The typical way to create a custom SymPy object is to subclass an existing
+SymPy class, generally either \texttt{Basic}, \texttt{Expr}, or
+\texttt{Function}. All SymPy classes should be subclasses of the base class
```

Comment:
All SymPy classes intended to be put into expression trees. Mention the polynomial module as a notable exception to this?